rclcpp/include/rclcpp/function_traits.hpp

```diff
@@ -69,6 +69,47 @@ struct arity_comparator
   static constexpr bool value = (Arity == function_traits<FunctorT>::arity);
 };
 
+template<bool same_arity, typename FunctorT, typename First, typename ... Last>
+struct check_argument_types_recursive
+{
+  static constexpr bool value = std::is_same<
+    typename function_traits<FunctorT>::template argument_type<
+      function_traits<FunctorT>::arity - sizeof ... (Last) -1
+    >,
+    First
+    >::value && check_argument_types_recursive<same_arity, FunctorT, Last ...>::value;
+};
+
+template<typename FunctorT, typename First, typename ... Last>
+struct check_argument_types_recursive<false, FunctorT, First, Last ...>
+{
+  static constexpr bool value = false;
+};
+
+template<bool same_arity, typename FunctorT, typename Arg>
+struct check_argument_types_recursive<same_arity, FunctorT, Arg>
+{
+  static constexpr bool value = std::is_same<
+    typename function_traits<FunctorT>::template argument_type<
+      function_traits<FunctorT>::arity - 1
+    >,
+    Arg
+    >::value;
+};
+
+template<typename FunctorT, typename Arg>
+struct check_argument_types_recursive<false, FunctorT, Arg>
+{
+  static constexpr bool value = false;
+};
+
+template<typename FunctorT, typename First, typename ... Last>
+struct check_argument_types
+{
+  static constexpr bool value = check_argument_types_recursive<arity_comparator<
+      sizeof ... (Last) +1, FunctorT>::value, FunctorT, First, Last ...>::value;
```

Comment:
The indentation looks a bit weird. Could the previous line be wrapped after the first `<`? Doesn't the `+1` in this line need a space between?